sql/lease.go

```diff
@@ -0,0 +1,546 @@
+	if err != nil {
+		panic(err) // Should never happen, we're careful to only encode supported datum types.
+	}
+	return k
+}
+
+func makeLeaseScanKeys(descID ID, version uint32, expiration int64) ([]byte, []byte) {
+	index := LeaseTable.PrimaryIndex
+
+	prefix := MakeIndexKeyPrefix(LeaseTable.ID, index.ID)
+	startKey := append([]byte(nil), prefix...)
+	startKey, err := encodeTableKey(startKey, parser.DInt(descID))
+	if err != nil {
+		panic(err) // Should never happen, we're careful to only encode supported datum types.
+	}
+	startKey, err = encodeTableKey(startKey, parser.DInt(version))
+	if err != nil {
+		panic(err) // Should never happen, we're careful to only encode supported datum types.
+	}
+	startKey, err = encodeTableKey(startKey, parser.DTimestamp{Time: nanosToTime(expiration)})
+	if err != nil {
+		panic(err) // Should never happen, we're careful to only encode supported datum types.
+	}
+
+	endKey := prefix
+	endKey, err = encodeTableKey(endKey, parser.DInt(descID))
+	if err != nil {
+		panic(err) // Should never happen, we're careful to only encode supported datum types.
+	}
+	endKey, err = encodeTableKey(endKey, parser.DInt(version)+1)
+	if err != nil {
+		panic(err) // Should never happen, we're careful to only encode supported datum types.
+	}
+	return startKey, endKey
+}
+
+// LeaseState holds the state for a lease. Exported only for testing.
+type LeaseState struct {
+	TableDescriptor
+	expiration int64
+	refcount   int
+}
+
+func (s *LeaseState) String() string {
+	return fmt.Sprintf("%d:%d", s.Version, s.expiration)
+}
+
+// Expiration returns the expiration time of the lease.
+func (s *LeaseState) Expiration() time.Time {
+	return nanosToTime(s.expiration)
+}
+
+// Refcount returns the reference count of the lease.
+func (s *LeaseState) Refcount() int {
+	return s.refcount
+}
+
+// LeaseStore implements the operations for acquiring and releasing leases and
+// publishing a new version of a descriptor. Exported only for testing.
+type LeaseStore struct {
+	db     client.DB
+	clock  *hlc.Clock
+	nodeID uint32
+}
+
+// jitteredLeaseDuration returns a randomly jittered duration from the interval
+// [0.75 * leaseDuration, 1.25 * leaseDuration].
+func (s LeaseStore) jitteredLeaseDuration() int64 {
+	return int64(float64(leaseDuration) * (0.75 + 0.5*rand.Float64()))
+}
+
+// Acquire a lease on the most recent version of a table descriptor.
+func (s LeaseStore) Acquire(tableID ID, minVersion uint32) (*LeaseState, error) {
+	lease := &LeaseState{}
+	lease.expiration = s.clock.Now().WallTime + s.jitteredLeaseDuration()
+	descKey := MakeDescMetadataKey(tableID)
+
+	err := s.db.Txn(func(txn *client.Txn) error {
+		if err := txn.GetProto(descKey, &lease.TableDescriptor); err != nil {
+			return err
+		}
+		if err := lease.Validate(); err != nil {
+			return err
+		}
+		if lease.Version < minVersion {
+			return fmt.Errorf("version %d of table %d does not exist yet", minVersion, tableID)
```

Comment:
No, it isn't expected in practice, but failure to do this check would cause an infinite loop if another bug requested a version to be acquired that had not been written. Caught by the tests. That's why we write them.